inst/@double/adjoint.m

```diff
@@ -0,0 +1,89 @@
+%% Copyright (C) 2022 Alex Vong
+%%
+%% This file is part of OctSymPy.
+%%
+%% OctSymPy is free software; you can redistribute it and/or modify
+%% it under the terms of the GNU General Public License as published
+%% by the Free Software Foundation; either version 3 of the License,
+%% or (at your option) any later version.
+%%
+%% This software is distributed in the hope that it will be useful,
+%% but WITHOUT ANY WARRANTY; without even the implied warranty
+%% of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+%% the GNU General Public License for more details.
+%%
+%% You should have received a copy of the GNU General Public
+%% License along with this software; see the file COPYING.
+%% If not, see <https://www.gnu.org/licenses/>.
+
+%% -*- texinfo -*-
+%% @documentencoding UTF-8
+%% @defun adjoint (@var{M})
+%% Numerical Classical Adjoint / Adjugate of a square matrix.
+%%
+%% @strong{Note on Terminology}: This function computes the
+%% ``Classical Adjoint'' / Adjugate of @var{M}. For the Conjugate Transpose /
+%% Hermitian adjoint (which is commonly referred to as the ``Adjoint'' in
+%% modern usage), @pxref{@@double/ctranspose}.
+%%
+%% Example:
+%% @example
+%% @group
+%% M = [-3 2 -5; -1 0 -2; 3 -4 1];
+%% A = adjoint (M)
+%%   @result{} A =
+%%        -8   18   -4
+%%        -5   12   -1
+%%         4   -6    2
+%% @end group
+%% @end example
+%%
+%% And note the following equalities involving the Classical Adjoint @code{A}:
+%% @example
+%% @group
+%% M * A
+%%   @result{} ans =
+%%       -6   0   0
+%%        0  -6   0
+%%        0   0  -6
+%% @end group
+%%
+%% @group
+%% A * M
+%%   @result{} ans =
+%%       -6   0   0
+%%        0  -6   0
+%%        0   0  -6
+%% @end group
+%%
+%% @group
+%% det (M) * eye (3)
+%%   @result{} ans =
+%%     Diagonal Matrix
+%%
+%%       -6   0   0
+%%        0  -6   0
+%%        0   0  -6
+%% @end group
+%% @end example
+%% @seealso{@@sym/adjoint}
+%% @end defun
+
+%% Reference: https://en.wikipedia.org/wiki/Adjugate_matrix
+
+
+function A = adjoint (M)
+  A = double (adjoint (vpa (M)));
+end
+
+
+%!test
+%! M = randn (2);
+%! A = [M(2,2) -M(1,2); -M(2,1) M(1,1)];
+%! assert (isequal (adjoint (M), A));
+%! assert (isequal (adjoint (adjoint (M)), M));
```

Comment:
these look like floating point equality tests.  Is it reasonable to expect a double-precision function to satisfy this test?

Reply:
Indeed, should I do `abs (x - y) < 1e-15` instead?

Reply:
I suppose so, in a relative sense would be better.  Maybe some of the other `@double/<fcn>` have some clues.

I worry a bit about the numerical stability of this sort of implementation (SymPy is not well-known for stability of its implementations when `sympy.Float` is involved: basically unless it calls `mpmath` I don't really trust it).

The short-term fix is that I think a random test is ill-advised unless you know the algorithm used is numerically stable.  I'd suggest a couple hardcoded cases.

Reply:
How about a nearly singular one?  `[0 1; 0.0001 1]`.  Does that still give 1e-15?

Reply:
It works for this particular case in sympy 1.10.1 with the default method of computing determinant. But let's keep it simple and use a fixed constant to avoid any future issues...